BulkSave existing documents doesn't apply save middleware hook updates

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

7.4.5

### Node.js version

18.16.0

### MongoDB server version

5.7

### Typescript version (if applicable)

5.1.6

### Description

When the bulkSave creates documents the modifications in the "save" middleware hook at persisted.
Subsequent updates made to the documents in "save" middleware hook are not persisted.  

Confirmed that the hook is being executed during the update.  See console.log below

### Steps to Reproduce

```typescript
describe("bulkSave middleware bug", async () => {
    it.only("it should be able to detect middleware save hook", async () => {
        const schema = new Schema({ name: String, _age: { type: Number, min: 0, default: 0 } })
        schema.pre("save", function () {
            this._age = this._age + 1
            console.log(`Updating person _age to ${this._age}`)
        })
        const connection = connectDatabase(mongodb.uri, "bulkSave")

        const Person = mongoose.model("Person", schema, "Persons", { connection })
        const person = new Person({ name: "Jean-Luc Picard", _age: 59 })

        await Person.bulkSave([person])

        let updatedPerson = await Person.findById(person._id)

        // this succeeds 
        expect(updatedPerson?._age).to.equal(60).

        await Person.bulkSave([updatedPerson!])

        updatedPerson = await Person.findById(person._id)
        
        // this fails - actual value is still 60
        expect(updatedPerson?._age).to.equal(61)
    })
})
```
Output

Updating person _age to 60
Updating person _age to 61
    1) it should be able to detect middleware save hook

  0 passing (113ms)
  1 failing

  1) bulkSave middleware bug
       it should be able to detect middleware save hook:

      AssertionError: expected 60 to equal 61
      + expected - actual

      -60
      +61
      
      at Context.<anonymous> (file:///Users/tate/code/Projects/HealthService/NodeServices/TypeScriptLibrary/src/__tests__/test.utils.ts:110:40)
      at processTicksAndRejections (node:internal/process/task_queues:95:5)




### Expected Behavior

Expect updates made in the save hook middleware to persist during a bulkSave.

--- a/lib/model.js
+++ b/lib/model.js
@@ -3565,10 +3565,8 @@ Model.bulkWrite = async function bulkWrite(ops, options) {
  * @param {Boolean} [options.j=true] If false, disable [journal acknowledgement](https://www.mongodb.com/docs/manual/reference/write-concern/#j-option)
  *
  */
-Model.bulkSave = async function(documents, options) {
+Model.bulkSave = async function bulkSave(documents, options) {
   options = options || {};
-
-  const writeOperations = this.buildBulkWriteOperations(documents, { skipValidation: true, timestamps: options.timestamps });
 
   if (options.timestamps != null) {
     for (const document of documents) {
@@ -3585,6 +3583,8 @@ Model.bulkSave = async function(documents, options) {
   }
 
   await Promise.all(documents.map(buildPreSavePromise));
+
+  const writeOperations = this.buildBulkWriteOperations(documents, { skipValidation: true, timestamps: options.timestamps });
 
   const { bulkWriteResult, bulkWriteError } = await this.bulkWrite(writeOperations, options).then(
     (res) => ({ bulkWriteResult: res, bulkWriteError: null }),

--- a/test/document.test.js
+++ b/test/document.test.js
@@ -12478,6 +12478,28 @@ describe('document', function() {
     await doc.save();
     assert.ok(doc);
   });
+
+  it('bulkSave() picks up changes in pre("save") middleware (gh-13799)', async() => {
+    const schema = new Schema({ name: String, _age: { type: Number, min: 0, default: 0 } });
+    schema.pre('save', function() {
+      this._age = this._age + 1;
+    });
+
+    const Person = db.model('Person', schema, 'Persons');
+    const person = new Person({ name: 'Jean-Luc Picard', _age: 59 });
+
+    await Person.bulkSave([person]);
+
+    let updatedPerson = await Person.findById(person._id);
+
+    assert.equal(updatedPerson?._age, 60);
+
+    await Person.bulkSave([updatedPerson]);
+
+    updatedPerson = await Person.findById(person._id);
+
+    assert.equal(updatedPerson?._age, 61);
+  });
 });
 
 describe('Check if instance function that is supplied in schema option is availabe', function() {

--- a/test/document.test.js
+++ b/test/document.test.js
@@ -12500,7 +12500,7 @@ describe('document', function() {
 
     assert.equal(updatedPerson?._age, 61);
   });
-    
+
   it('handles default embedded discriminator values (gh-13835)', async function() {
     const childAbstractSchema = new Schema(
       { kind: { type: Schema.Types.String, enum: ['concreteKind'], required: true, default: 'concreteKind' } },

